pkg/generate/code/set_resource.go

```diff
@@ -770,6 +769,206 @@ func SetResourceGetAttributes(
+// r.ko.Spec.ResourceID = identifier.NameOrID
+//
+// f0, f0ok := identifier.AdditionalKeys["scalableDimension"]
+// if f0ok {
+// 	  r.ko.Spec.ScalableDimension = f0
+// }
+// f1, f1ok := identifier.AdditionalKeys["serviceNamespace"]
+// if f1ok {
+// 	  r.ko.Spec.ServiceNamespace = f1
+// }
+// ```
+
+func SetResourceIdentifiers(
+	cfg *ackgenconfig.Config,
+	r *model.CRD,
+	// String representing the name of the variable that we will grab the Input
+	// shape from. This will likely be "identifier" since in the templates that
+	// call this method, the "source variable" is the CRD struct which is used
+	// to populate the target variable, which is the struct of unique
+	// identifiers
+	sourceVarName string,
+	// String representing the name of the variable that we will be **setting**
+	// with values we get from the Output shape. This will likely be
+	// "r.ko" since that is the name of the "target variable" that the
+	// templates that call this method use for the Input shape.
+	targetVarName string,
+	// Number of levels of indentation to use
+	indentLevel int,
+) string {
+	op := r.Ops.ReadOne
+	if op == nil {
+		// If single lookups can only be done using ReadMany
+		op = r.Ops.ReadMany
+	}
+	inputShape := op.InputRef.Shape
+	if inputShape == nil {
+		return ""
+	}
+
+	primaryKeyOut := "\n"
+	arnOut := ""
+	additionalKeyOut := "\n"
+
+	indent := strings.Repeat("\t", indentLevel)
+
+	primaryKeyOut += fmt.Sprintf("%sif %s.NameOrID == nil {\n", indent, sourceVarName)
+	primaryKeyOut += fmt.Sprintf("%s\treturn ackerrors.MissingNameIdentifier\n", indent)
+	primaryKeyOut += fmt.Sprintf("%s}\n", indent)
+
+	primaryIdentifier := ""
+
+	// Attempt to fetch the primary identifier from the configuration
+	opConfig, ok := cfg.Operations[op.Name]
+	if ok {
+		primaryIdentifier = opConfig.PrimaryIdentifierFieldName
+	}
+
+	// Determine the "primary identifier" based on the names of each field
+	if primaryIdentifier == "" {
+		primaryIdentifierLookup := []string{
+			"Name",
+			r.Names.Original + "Name",
+			r.Names.Original + "Id",
+		}
+
+		for _, memberName := range inputShape.MemberNames() {
+			if util.InStrings(memberName, primaryIdentifierLookup) {
+				if primaryIdentifier == "" {
+					primaryIdentifier = memberName
+				} else {
+					panic("Found multiple possible primary identifiers for " +
+						r.Names.Original + ". Set " +
+						"`primary_identifier_field_name` for the " + op.Name +
+						" operation in the generator config.")
+				}
+			}
+		}
+
+		// Still haven't determined the identifier? Panic
+		if primaryIdentifier == "" {
+			panic("Could not find primary identifier for " + r.Names.Original +
+				". Set `primary_identifier_field_name` for the " + op.Name +
+				" operation in the generator config.")
+		}
+	}
+
+	paginatorFieldLookup := []string{
+		"NextToken",
+		"MaxResults",
+	}
+
+	additionalKeyCount := 0
+	for _, memberName := range inputShape.MemberNames() {
+		if r.UnpacksAttributesMap() && memberName == "Attributes" {
```

Comment:
Note that `r.UnpacksAttributesMap()` will never be true if the `op` is not `GetAttributes`. I'd recommend just erroring out when `op` is `GetAttributes` because none of this logic will work :)